internal/messaging/response_handler.go

```diff
@@ -1,147 +1,536 @@
 	"go.uber.org/zap"
 )
 
-var _ ResponseHandler = (*TvmResponseHandler)(nil)
+var _ ResponseHandler = (*EvmResponseHandler)(nil)
 
 type ResponseHandler interface {
 	HandleResponse(ctx context.Context, msgType MessageType, request *RequestContent, response *ResponseContent)
+	HandleRequest(ctx context.Context, msgType MessageType, request *RequestContent) error
 }
-type TvmResponseHandler struct {
-	tvmClient *tvm.Client
+type EvmResponseHandler struct {
+	ethClient *ethclient.Client
 	logger    *zap.SugaredLogger
+	pk        *secp256k1.PrivateKey
+	cfg       *config.EvmConfig
 }
 
-func (h *TvmResponseHandler) HandleResponse(ctx context.Context, msgType MessageType, request *RequestContent, response *ResponseContent) {
+func (h *EvmResponseHandler) HandleResponse(ctx context.Context, msgType MessageType, request *RequestContent, response *ResponseContent) {
 	switch msgType {
 	case MintRequest: // distributor will post-process a mint request to buy the returned NFT
 		if h.handleMintRequest(ctx, response) {
 			return
 		}
-	case MintResponse: // provider will act upon receiving a mint response by minting an NFT
+	case MintResponse: // supplier will act upon receiving a mint response by minting an NFT
 		if h.handleMintResponse(ctx, response, request) {
 			return
 		}
 	}
 }
 
-func (h *TvmResponseHandler) handleMintResponse(ctx context.Context, response *ResponseContent, request *RequestContent) bool {
-	owner := h.tvmClient.Address()
+func (h *EvmResponseHandler) HandleRequest(_ context.Context, msgType MessageType, request *RequestContent) error {
+	switch msgType { //nolint:gocritic
+	case MintRequest:
+		request.BuyerAddress = crypto.PubkeyToAddress(h.pk.ToECDSA().PublicKey).Hex()
+	}
+	return nil
+}
+
+func (h *EvmResponseHandler) handleMintResponse(_ context.Context, response *ResponseContent, request *RequestContent) bool {
 	if response.MintResponse.Header == nil {
 		response.MintResponse.Header = &typesv1alpha.ResponseHeader{}
 	}
-	buyer, err := codec.ParseAddressBech32(consts.HRP, request.MintRequest.BuyerAddress)
+	bookingTokenABIFile := h.cfg.BookingTokenABIFile
+	abi, err := loadABI(bookingTokenABIFile)
+	if err != nil {
+		addErrorToResponseHeader(response, fmt.Sprintf("error loading ABI: %v", err))
+		return true
+	}
+	address := crypto.PubkeyToAddress(h.pk.ToECDSA().PublicKey)
+
+	bookingTokenAddress := common.HexToAddress(h.cfg.BookingTokenAddress)
+
+	packedData, err := abi.Pack("getSupplierName", address)
+	if err != nil {
+		h.logger.Infof("Error packing data for getSupplierName: %v", err)
+		addErrorToResponseHeader(response, fmt.Sprintf("error packing data: %v", err))
+		return true
+	}
+
+	msg := ethereum.CallMsg{
+		To:   &bookingTokenAddress,
+		Data: packedData,
+	}
+	result, err := h.ethClient.CallContract(context.Background(), msg, nil)
 	if err != nil {
-		addErrorToResponseHeader(response, fmt.Sprintf("error parsing buyer address: %v", err))
+		h.logger.Infof("Error calling contract: %v", err)
```

Comment:
I would also suggest instead of defining error text twice, do something like:
```go
err = fmt.Errorf("error calling contract: %w", err)
h.logger.Info(err)
addErrorToResponseHeader(response, err.String())
```
Sure, its one more line, but makes it easier to handle changes
See similar places